sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamTableFunctionScanRel.java

```diff
@@ -232,7 +233,11 @@ private Schema getKeySchema(Schema inputSchema, List<Integer> keys) {
           upstream
               .apply(
                   "assignEventTimestamp",
-                  WithTimestamps.<Row>of(row -> row.getDateTime(windowFieldIndex).toInstant())
+                  WithTimestamps.<Row>of(
+                          row ->
+                              Instant.ofEpochMilli(
```

Comment:
It seems concerning that this is silently lossy...

Reply:
We plan to support windowing only up to millisecond precision, and the timestamp here seems only used by windowing. So I guess it is fine?

Reply:
Yeah. We do not support timestamp microsecond precision for windowing in this PR. Windowing for Streaming tests is still millisecond precision. So this conversion is fine, because `Instant.toEpochMilli()` is not lossy for an millisecond precision `Instant`.

Reply:
I think there'll be surprising loss of precision if this field is used for windowing (say, with TimestampCombiner.EARLIEST) and then the timestamp is inspected (is even available from SQL, right?) and has been truncated. (Or any other conversion from Row to Beam and back again being lossy.)

Reply:
One option would be if one is forced to coerce to a millisecond-granularity one before using it in windowing (or other Beam-timestamp-using) operations.

Reply:
> I think there'll be surprising loss of precision if this field is used for windowing (say, with TimestampCombiner.EARLIEST) and then the timestamp is inspected (is even available from SQL, right?) and has been truncated.

I see. From Java user can definitely inspect the timestamp, so that is a problem. From pure SQL (e.g. from Dataflow SQL UI) I think for TUMBLE and HOP this will be fine, because the window start/end timestamp will not have a microsecond component if we only allow windowing up to millisecond. For SESSION I think there will be a problem, because the window starts/ends at a timestamp of an element.

I will make sure we don't silently truncate precision before I merge the PR. One option is to use a PipelineOptions for user to choose truncate or fail over timestamp with microsecond component during windowing.

Reply:
That would be great. We could put this as an option on the SQLTransform itself.

Reply:
+1 for an option on the SQLTransform itself, that will be easier to use from Python